GCRCatalogs/register.py

```diff
@@ -1,301 +1,492 @@
-    return d
+                        self.resolve_root_dir(c, config_name, record)
+
+        return config_dict
 
 
-class Config():
-    def __init__(self, config_path, config_dir=''):
+class Config(Mapping):
+    YAML_EXTENSIONS = (".yaml", ".yml")
+
+    def __init__(self, config_path, config_dir="", resolvers=None):
         self.path = os.path.join(config_dir, config_path)
         self.basename = os.path.basename(self.path)
         self.rootname, self.ext = os.path.splitext(self.basename)
         self.name = self.rootname.lower()
-        self._content = None
+
+        self._resolvers = None
+        self._content_ = None
+        self._resolved_content_ = None
+
+        if resolvers:
+            self.set_resolvers(*resolvers)
+
+    def set_resolvers(self, *resolvers):
+        if not all(map(callable, resolvers)):
+            raise ValueError("`resolvers` should be callable.")
+        self._resolvers = resolvers
+        self.reset_resolved_content()
 
     @property
     def ignore(self):
-        return (
-            self.rootname.startswith('_') or
-            self.ext.lower() not in _YAML_EXTENSIONS
-        )
+        return self.rootname.startswith("_") or self.ext.lower() not in self.YAML_EXTENSIONS
+
+    @property
+    def _content(self):
+        if self._content_ is None:
+            self._content_ = load_yaml_local(self.path)
+        return self._content_
+
+    @property
+    def _resolved_content(self):
+        if self._resolved_content_ is None:
+            content = self.content
+            if self._resolvers:
+                for resolver in self._resolvers:
+                    content = resolver(content, self.name)
+            if not self.is_pseudo and "subclass_name" not in content:
+                raise ValueError(
+                    "`subclass_name` is missing in the config of {}"
+                    "and all of its references".format(self.name)
+                )
+            self._resolved_content_ = content
+        return self._resolved_content_
+
+    def reset_resolved_content(self):
+        self._resolved_content_ = None
 
     @property
     def content(self):
-        if self._content is None:
-            self._content = load_yaml_local(self.path)
-        return self._content
-    
-class ConfigRegister():
+        return copy.deepcopy(self._content)
+
+    @property
+    def resolved_content(self):
+        return copy.deepcopy(self._resolved_content)
+
+    def __getitem__(self, key):
+        return self._content[key]
+
+    def __iter__(self):
+        return iter(self._content)
+
+    def __len__(self):
+        return len(self._content)
+
+    @property
+    def is_default(self):
+        return self.get("include_in_default_catalog_list")
+
+    @property
+    def is_pseudo(self):
+        return self.get("is_pseudo_entry")
+
+    @property
+    def is_alias(self):
+        return self.get("alias")
+
+    @staticmethod
+    def _has_reference_keys(d):
```

Comment:
Suggest you add a static to the class, e.g. `REF_KEYS = ("alias", "based_on")` since they appear in two or three different places in the code, and it's conceivable at some point we might add another reference key needing resolution.

Reply:
Sure. While I am at this, I made all special key names static attribute to `Config`.